src/impl_arc_array.rs

```diff
@@ -0,0 +1,23 @@
+// Copyright 2019 ndarray developers.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use alloc::sync::Arc;
+use crate::imp_prelude::*;
+
+/// Methods specific to `ArcArray`.
+///
+/// ***See also all methods for [`ArrayBase`]***
+impl<A, D> ArcArray<A, D>
+where
+    D: Dimension,
+{
+    /// Returns `true` iff the inner `Arc` is not shared.
+    pub fn is_unique(&mut self) -> bool {
```

Comment:
Is it possible to make this a non-mut method?

Reply:
Wouldn't it risk creating a race condition? I kinda got confused on this point... The alternative would be just to sum strong and weak counts:
```
fn is_unique(&self) -> bool {
    Arc::strong_count(&self.data.0) + Arc::weak_count(&self.data.0) == 1
}
```

Reply:
There are race conditions anyway, because the method only returns a boolean. When is_unique has finished and returns its value, it could already be shared again. There's no solving that without creating something entirely different.

What's the use case for this?

Reply:
In my case I am using an `ArcArray` as the core of a struct that I call from other language. I want the user of that language to be able to know when a certain method, lets say for example fft calculation, has created a copy of that inner `ArcArray`.

Reply:
It is confusing, and if there would be weak refs you're right that it's racy. ArcArray doesn't support weak refs at all. That means there are two options - (1) the way of just checking strong count and deciding we can never support Weak refs or (2) your get_mut implementation which is more sure to always be safe.

Any "is_unique -> bool" method is racy unless we can control the current Arc, i.e we are sure nobody's cloning the exact Arc handle we are holding. For this reason, ownership or &mut of the current Arc is needed to be sure of that.

I would say both (1) in combination with holding or having the option to hold a &mut Arc would be just as safe as (2) for ArcArray, but which choice to use doesn't matter so much to me.


If there is code that first .clone() and then mutate (unsharing) your ArcArray, you won't notice this way, because the final effect is two independent copies of the data.

Reply:
You are saying something like this for (1)?
```
fn is_unique(&mut self) -> bool {
    Arc::strong_count(&self.0) == 1
}
```

I am ok with both choices, I just used the second one because of extensibility, in case you choose to use weak references in the future, for whatever reason. Let me know if you have any better suggestion than this.

Reply:
I think it's clear it's possible to make it a non-mut method - using strong_count.
That seems to be compatible with the use case as well (your use case - if you want to ensure the Arc is not concurrently cloned, you can hold a &mut to the ArcArray, the method doesn't need to use &mut for that). I can't imagine too many other use cases, other than some user wanting to do something they shouldn't w.r.t mutability and sharing :laughing:.

Any other ideas about use cases?
Maybe get_mut is the safest choice for unknown use cases?

Reply:
I will modify it then to count strong count and accept self. Another approach would be to allow access (a getter) to the inner arc, then the user could do whatever he wanted with it.
Just for curiosity, why do you want to make it non-mut?

Reply:
For non-mut, just the instinct that it's an interrogating method that doesn't modify.
Inner fields won't be exposed :slightly_smiling_face: